xiilib/databases.py

```diff
@@ -54,16 +56,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str):
             charm: The requesting charm object.
             relation_name: The redis relation name.
         """
-        self._charm = charm
-        self.relation_name = relation_name
-        # redis charm doesn't provide database id via the relation
+        # we don't need store for Redis
+        self._redis = RedisRequires(charm=charm, _stored=MagicMock(), relation_name=relation_name)
         self.database = ""
-        self.on = self._RedisDatabaseRequiresEventShim(
-            self._charm.on[relation_name].relation_changed
-        )
+        self.relation_name = relation_name
+        relation = charm.model.get_relation(relation_name)
+        self.relation_id = relation.id if relation else None
+        self.on = self._RedisDatabaseRequiresEventShim(charm.on.redis_relation_updated)
```

Comment:
If we keep the Redis handling within the database module, it can share all the database handling logic inside the same module. For example, check if the database is declared in the metadata, block operations if a non-optional database is not ready, generate the database environment variables, and use the same database event naming convention.